{
  "cells": [
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "title: \"Use GenomeDK\"\n",
        "subtitle: \"An introduction to the GDK system and basic commands https://hds-sandbox.github.io/GDKworkshops\"\n",
        "author: \n",
        "    - \"Samuele Soraggi\"\n",
        "    - \"Dan SÃ¸ndergaard\"\n",
        "institute:\n",
        "    - Health Data Science sandbox, BiRC\n",
        "    - GenomeDK, Health\n",
        "date: last-modified\n",
        "format: revealjs\n",
        "toc: true\n",
        "toc-depth: 1\n",
        "slide-number: h.v\n",
        "code-line-numbers: false\n",
        "logo: img/logos.png\n",
        "navigation-mode: vertical\n",
        "---"
      ],
      "id": "79dc183d"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# GenomeDK's ABC\n",
        "\n",
        "Learn your way around the basics of the `GenomeDK` cluster.\n",
        "\n",
        "## Infrastructure\n",
        "\n",
        "`GenomeDK` is a **computing cluster**, i.e. a set of interconnected computers (nodes). `GenomeDK` has:\n",
        "\n",
        "- **computing nodes** used for running programs (~15000 cores)\n",
        "- **storage nodes** storing data in many hard drives (~23 PiB)\n",
        "- a **network** making nodes talk to each other\n",
        "- a **frontend** node from which you can send your programs to a node to be executed\n",
        "- a **queueing system** called *slurm* to prioritize the users' program to be run"
      ],
      "id": "c930feb6"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "![](./img/Main-components-of-a-computer-cluster.png){fig-align=\"center\"}\n",
        "\n",
        "## Access\n",
        "\n",
        "- **Creating an account** happens through [this form](https://console.genome.au.dk/user-requests/create/) at genome.au.dk\n",
        "\n",
        "    ![](img/account.png){width=600px fig-alig=\"center\"}\n",
        "\n",
        "---"
      ],
      "id": "86482a34"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "- **Logging into GenomeDK** happens through the command ^[both in Linux/Mac terminal and Windows Powershell. Powershell might require `ssh.exe` instead of `ssh`]\n",
        "\n",
        "    ```{.bash}\n",
        "    [local]$  ssh USERNAME@login.genome.au.dk\n",
        "    ```\n",
        "\n",
        "- When first logged in, **setup the 2-factor authentication** by \n",
        "    - showing a QR-code with the command\n",
        "\n",
        "        ```{.bash}\n",
        "        [fe-open-01]$  gdk-auth-show-qr\n",
        "        ```\n",
        "    - scanning it with your phone's Authenticator app ^[e.g. Microsoft Authenticator, Google Authenticator, ...].\n",
        "\n",
        "\n",
        "## Access without password\n",
        "\n",
        "It is nice to avoid writing the password at every access. If you are on the cluster, exit from it to go back to your local computer\n",
        "\n",
        "```{.bash}\n",
        "exit\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Now we generate an RSA key.\n",
        "\n",
        "```{.bash}\n",
        "ssh-keygen -t ed25519\n",
        "```\n",
        "\n",
        "Always press <kbd>Enter</kbd> and do not insert any password when asked. "
      ],
      "id": "f1f58f45"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        ":::: {.columns}\n",
        "\n",
        "::: {.column width=\"50%\" }\n",
        "\n",
        "Wait a second! What is an RSA key?\n",
        "\n",
        "1. it's a pair of long codes, called *public* and *private*\n",
        "2. the two keys are *mathematically related*\n",
        "3. the user has  *only private key*\n",
        "4. the cluster has *only public key*\n",
        "5. the user asks to be authenticated to GenomeDK (AuthReq)\n",
        ":::\n",
        "\n",
        "::: {.column width=\"50%\"}\n",
        "```\n",
        "+--------+         +--------+\n",
        "|  User  |         |GenomeDK|\n",
        "|        |         |        |\n",
        "| PrivKey| ------> | PubKey |\n",
        "|(id_rsa)| AuthReq |(auth)  |\n",
        "+--------+         +--------+\n",
        "     |                   ^\n",
        "     v                   |\n",
        "  SignReq             VerifyReq\n",
        "     |                   |\n",
        "     v                   v\n",
        "  Access Granted if Verified\n",
        "\n",
        "```\n",
        "\n",
        "6. the user *signs the request* with private key (SignReq)\n",
        "7. GenomeDK *verifies* the request through match with the public key (VerifyReq)\n",
        "8. If verification succeeds, access is granted\n",
        "\n",
        ":::\n",
        "\n",
        "::::\n",
        "\n",
        "---"
      ],
      "id": "1c288866"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "We create a folder on the cluster called `.ssh` to contain the RSA key we created\n",
        "\n",
        "```{.bash}\n",
        "ssh <USERNAME>@login.genome.au.dk mkdir -p .ssh\n",
        "```\n",
        "&nbsp;\n",
        "\n",
        "and finally send the RSA public key to the cluster, into the file `authorized_keys`\n",
        "\n",
        "```{.bash}\n",
        "cat ~/.ssh/id_rsa.pub | ssh username@login.genome.au.dk 'cat >> .ssh/authorized_keys'\n",
        "```\n",
        "&nbsp;\n",
        "\n",
        "this time will be the last where you are asked your password when logging in from your computer.\n",
        "\n",
        "# File System (FS) on GenomeDK\n",
        "\n",
        "Directory structure and how to navigate it\n",
        "\n",
        "## How is the FS organized\n",
        "\n",
        ":::: {.columns}\n",
        "\n",
        "::: {.column width=\"50%\" }\n",
        "\n",
        "Folders and files follow a tree-like structure\n",
        "\n",
        "- `/` is the root of the filesystem - nothing is above that\n",
        "- `home` and `faststorage` are two **root folders**\n",
        "- projects are in `faststorage` and **linked** to your home\n",
        "- you can reach files and folders with a **path**\n",
        "\n",
        ":::\n",
        "\n",
        "::: {.column width=\"50%\"}\n",
        "\n",
        "![](img/complexTree.png){fig-align=\"center\"}\n",
        "\n",
        ":::\n",
        "\n",
        "::::"
      ],
      "id": "bdb87774"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        ":::: {.columns}\n",
        "\n",
        "::: {.column width=\"50%\" }\n"
      ],
      "id": "3ee07c35"
    },
    {
      "cell_type": "code",
      "metadata": {},
      "source": [
        "#| echo: false\n",
        "from jupyterquiz import display_quiz\n",
        "git_url='https://raw.githubusercontent.com/hds-sandbox/scverse-2024-workshop/main/Questions/quiz00.json'\n",
        "display_quiz(git_url, 1, shuffle_answers=True)"
      ],
      "id": "e5c99632",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        ":::\n",
        "\n",
        "::: {.column width=\"50%\"}\n",
        "\n",
        "![](img/complexTree.png){fig-align=\"center\"}\n",
        "\n",
        ":::\n",
        "\n",
        "::::\n",
        "\n",
        "## Home on the cluster\n",
        "\n",
        "Log in again! Use <kbd>up</kbd> on the terminal to find the command you used previously\n",
        "\n",
        "Every time you log in, you will find yourself into your private home folder. This is denoted by `~` or equivalently `/home/username/`. Your prompt will show something like this:\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "```{.bash}\n",
        "[samuele@fe-open-02 ~]\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "which follows the format \n",
        "\n",
        "```{.bash}\n",
        "[username@node current_folder]\n",
        "```\n",
        "\n",
        "---"
      ],
      "id": "9392c0eb"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "See the content of the current folder with \n",
        "\n",
        "```{.bash}\n",
        "ls\n",
        "```\n",
        "\n",
        "or to see more details\n",
        "\n",
        "```{.bash}\n",
        "ls -lh\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        ":::{.callout-warning}\n",
        "do not fill up your home with data. It has a **limited amount of storage** (a quota of 100GB).\n",
        ":::\n",
        "\n",
        "# Project management\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "It is easy to start creating files everywhere in your project folders. Data, analysis files, results and the like. \n",
        "\n",
        "&nbsp;\n",
        "\n",
        "**Managing your folders rationally** is the best way of finding your way around. Especially when getting back to your analysis after long time.\n",
        "\n",
        "## Creation\n",
        "\n",
        "You need a project from which you can run your programs. Request a project with the command\n",
        "\n",
        "```{.bash}\n",
        "gdk-project-request -g <project_name>\n",
        "```\n",
        "\n",
        "This creates a folder with the desired name. You should be able to go into that folder:\n",
        "\n",
        "```{.bash}\n",
        "cd <project_name>\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "You can see how many resources your projects are using with \n",
        "\n",
        "```{.bash}\n",
        "gdk-project-usage\n",
        "```"
      ],
      "id": "2b20ff11"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "## Users management\n",
        "\n",
        "**Only the creator** (owner) can see the project folder. You can add an user\n",
        "\n",
        "```{.bash}\n",
        "gdk-project-add-user -g <project name> -u <username>\n",
        "```\n",
        "\n",
        "or remove it\n",
        "\n",
        "```{.bash}\n",
        "gdk-project-remove-user -g <project name> -u <username>\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "More about user's management [in the documentation](https://genome.au.dk/docs/projects-and-accounting/#managing-a-project)\n",
        " \n",
        "---"
      ],
      "id": "2cd1c2fa"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Folders management\n",
        "\n",
        "It is important to\n",
        "\n",
        "- have a coherent folder structure\n",
        "- backup only important things (raw data, analysis scripts)\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Remember: Storage cost >> Computation cost"
      ],
      "id": "0c2fa2e8"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "Example of structure, which backs up raw data and analysis\n",
        "\n",
        "![](img/structure.png)\n",
        "\n",
        "You can do it with a script:\n",
        "```{.bash}\n",
        "wget https://raw.githubusercontent.com/hds-sandbox/GDKworkshops/5bbfc11e3796d5f4f1af39aecd6858721aca1612/Scripts/populate_project.sh\n",
        "bash populate_project.sh\n",
        "```\n",
        "\n",
        "---"
      ],
      "id": "079dbb11"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "If your project has many users, a good structure can be\n",
        "\n",
        "![](img/multistructure.png)\n",
        "\n",
        "&nbsp; \n",
        "\n",
        "```{.bash}\n",
        "mkdir -p Backup/Data Workspaces/username1 Workspaces/username2\n",
        "ln -s Backup/Data/ Data\n",
        "```"
      ],
      "id": "fcb8b8ac"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "Each user can go in its folder inside the project and run the script to populate the folders\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "```{.bash}\n",
        "cd Workspaces/username1\n",
        "wget https://github.com/hds-sandbox/GDKworkshops/blob/be7315365c152ecd75e94c1f56e1578062c2c096/Scripts/populate_project.sh\n",
        "chmod +x populate_project.sh\n",
        "./populate_project.sh\n",
        "```\n",
        "\n",
        "# Downloads and Copies\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "In your daily life on a cluster you are going to need **downloads and exchange of files** with online archives and your local PC.\n",
        "\n",
        ":::{.callout-warning}\n",
        "\n",
        "Downloads should **always** happens on the `front-end` nodes, and never using a compute node when working on `GenomeDK`\n",
        "\n",
        ":::\n",
        "\n",
        "## Download with wget\n",
        "\n",
        "`wget` is a utility for **command-line-based downloads**. It is already installed on `GenomeDK` and works with `http`, `https`, `ftp` protocols.\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Example:\n",
        "\n",
        "```{.bash}\n",
        "wget -O ./output.png \\\n",
        "     -c \\\n",
        "     -b \\\n",
        "     https://example.com/image.png\n",
        "```\n",
        "downloads a `png` file and saves it as `output.png` (option `O`), downloads in background (`-b`) and if the download was interrupted earlier, it retrieves it from where it stopped (`-c`).\n",
        "\n",
        "---"
      ],
      "id": "bbb06738"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "`wget` has many options you can use, but what shown in the example above is what you need most times. You can see them with the command\n",
        "\n",
        "```{.bash}\n",
        "wget --help\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Also, you can find [this cheatsheet](https://opensource.com/sites/default/files/gated-content/osdc_cheatsheet-wget-2021.10.21.pdf) useful for remembering the commands to most of the things you can think about downloading files using wget. [At this page](https://opensource.com/article/21/10/linux-wget-command) there are also some concrete examples for `wget`.\n",
        "\n",
        "## SCP transfer\n",
        "\n",
        "`SCP` (Secure Copy Protocol) can transfer files securely\n",
        "\n",
        "- between a LOCAL and a REMOTE host\n",
        "- between TWO REMOTE hosts\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "You can use it to transfer files **from your pc to GenomeDK and viceversa**, but also **between GenomeDK and another computing cluster** (for example, downloading data from a collaborator, which resides on a different remote computing system)."
      ],
      "id": "e46c5ddf"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "To copy a file to GenomeDK from your local computer:\n",
        "```{.bash}\n",
        "scp /home/my_laptop/Documents/file.txt \\\n",
        "    username@login.genome.au.dk:/home/username/my_project/\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "The inverse operation just changes the order of the sender and receiver:\n",
        "```{.bash}\n",
        "scp username@login.genome.au.dk:/home/username/my_project/file.txt \\\n",
        "    /home/my_laptop/Documents/\n",
        "```\n",
        "\n",
        "---"
      ],
      "id": "1c8ba8ef"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "If you want to copy an entire folder, use the option -r (recursive copy). The previous examples become\n",
        "\n",
        "```{.bash}\n",
        "scp -r /home/my_laptop/Documents/folder \\\n",
        "       username@login.genome.au.dk:/home/username/my_project/\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "and\n",
        "\n",
        "```{.bash}\n",
        "scp -r username@login.genome.au.dk:/home/username/my_project/folder \\\n",
        "       /home/my_laptop/Documents/\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "A few more options are available and you can see them with the command ``` scp --help ```.\n",
        "\n",
        "## Rsync transfer\n",
        "\n",
        "Differently from `scp`, you can use `rsync` to **syncronize files and folders between two locations**. It copies only the changes in the data and not all of it every time.\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Copying a file or a folder between your computer and `GenomeDK` works exactly as in `scp`. For example\n",
        "\n",
        "```{.bash}\n",
        "rsync --progress -r \\\n",
        "      username@login.genome.au.dk:/home/username/my_project/folder \\\n",
        "      /home/my_laptop/Documents/\n",
        "```\n",
        "\n",
        "where we add an option to show a progress bar"
      ],
      "id": "c30a6532"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "An interrupted syncronization can be retrieved if interrupted. **To allow future retrieval of partial transfer**, the previous command needs the additional option `--partial` (which keeps partial downloads without deleting them):\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "```{.bash}\n",
        "rsync --partial --progress -r \\\n",
        "      username@login.genome.au.dk:/home/username/my_project/folder \\\n",
        "      /home/my_laptop/Documents/\n",
        "```\n",
        "&nbsp;\n",
        "\n",
        "After an interruption, just rerun the exact same command to retrieve the syncronization.\n",
        "\n",
        "If you have **large files**, the option `-z` (compression) reduces the amount of data (and the time) to transfer.\n",
        "\n",
        "## Interactive transfer\n",
        "\n",
        "You can also do transfering with an interactive software, such as `Filezilla`, which  has an easy interface. [Download Filezilla](https://filezilla-project.org/download.php?type=client).\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "When done, open `Filezilla` and use the following information on the login bar:\n",
        "\n",
        "- Host: `login.genome.au.dk`\n",
        "- Username, Password: your `GenomeDK` username and password\n",
        "- Port: `22`\n",
        "\n",
        "---"
      ],
      "id": "b0994ed2"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Press on `Quick Connect`. As a result, you will establish a secure connection to `GenomeDK`. On the left-side browser you can see your local folders and files. On the right-side, the folders and files on `GenomeDK` starting from your `home`.\\\n",
        "\n",
        "![](./img/filezilla1.png){fig-align=\"center\"}"
      ],
      "id": "4c5e31a2"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "If you right-click on any local file or folder, you can **upload** it immediately, or **add it to the transfer queue**. The file will end up in the selected folder of the right-side browser.\n",
        "\n",
        "![](./img/filezilla2.png){fig-align=\"center\"}\n",
        "\n",
        "---"
      ],
      "id": "c6bba808"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The **download** process works similarly using the right-side browser and choosing the destination folder on the left-side browser.\n",
        "\n",
        "![](./img/filezilla3.png){fig-align=\"center\"}"
      ],
      "id": "24509b13"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "If you have created a queue, this will be shown at the bottom of the window as a list. You can inspect destination folders from there and choose other options such as transfer priority.\n",
        "\n",
        "![](./img/filezilla4.png){fig-align=\"center\"}\n",
        "\n",
        "To start a queue, use `CTRL + P`, `Transfer --> Process Queue` or press  the button ![](./img/filezilla5.png){width=100} on the toolbar.\n",
        "\n",
        "# Package/Environment management\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Properly managing your **software and its dependencies** is fundamental for reproducibility\n",
        "\n",
        "## Virtual environments\n",
        "\n",
        "Each project needs specific software versions dependent on each other for **reproducibility** - without interferring with other projects.\n",
        "\n",
        "&nbsp;\n",
        "\n",
        ":::{.callout-note title=\"Definition\"}\n",
        "A **virtual environment** keeps project-specific softwares and their dependencies separated \n",
        "\n",
        "&nbsp;\n",
        "\n",
        "A **package manager** is a software that can retrieve, download, install, upgrade packages **easily and reliably**\n",
        ":::\n",
        "\n",
        "---"
      ],
      "id": "0d895801"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "![](./img/virtualenvs.png)\n",
        "\n",
        "## Conda\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Conda is **both** a virtual environment and a package manager.\n",
        "\n",
        "- easy to use and understand\n",
        "- can handle quite big environments\n",
        "- environments are easily shareable\n",
        "- a [large archive](https://anaconda.org) (Anaconda) of packages\n",
        "- active community of people archiving their packages on Anaconda\n",
        "\n",
        "## Installation\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Just download and execute the installer by\n",
        "```{.bash}\n",
        "wget https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh -O miniforge.sh\n",
        "chmod +x miniforge.sh\n",
        "bash miniforge.sh -b\n",
        "./miniforge3/bin/conda init bash\n",
        "```\n",
        "\n",
        "After a few `ENTER`s and `YES`'s you should get the installation done. Run\n",
        "\n",
        "```{.bash}\n",
        "source ~/.bashrc\n",
        "```\n",
        "\n",
        "and doublecheck that `conda` works:\n",
        "\n",
        "```{.bash}\n",
        "conda info\n",
        "```\n"
      ],
      "id": "f8fe84af"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "\n",
        "## Conda Configuration\n",
        "\n",
        "You can add some default *channels* where to find archived packages. Here are some tipycal ones\n",
        "\n",
        "```{.bash}\n",
        "conda config --append channels bioconda\n",
        "conda config --append channels genomedk\n",
        "conda config --append channels r\n",
        "conda config --append channels conda-forge\n",
        "```\n",
        "\n",
        "We tell `conda` to look into channels in the order specified above. We also avoid opening the `base` environment (where `conda` is installed) at login.\n",
        "\n",
        "```{.bash}\n",
        "conda config --set channel_priority strict\n",
        "conda config --set auto_activate_base false\n",
        "```\n",
        "\n",
        "## Base environment\n",
        "\n",
        "is the environment containing conda itself. The current environment is in your prompt, but you will not see it again after disabling it at login.\n",
        "\n",
        "```{.bash}\n",
        "(base) [samuele@fe-open-02 ~]$\n",
        "```\n",
        "\n",
        "We update Conda with `libmamba solver` - a lot faster in installing many packages at once.\n",
        "\n",
        "```{.bash}\n",
        "conda install -n base --yes conda-libmamba-solver\n",
        "conda config --set solver libmamba\n",
        "```\n",
        ":::{.callout-warning title=\"Don't touch the Base\"}\n",
        "This is **the only time** you should install in the `base` environment! You might otherwise ruin the conda installation.\n",
        ":::\n",
        "\n",
        "---"
      ],
      "id": "b77e2ca6"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Look at the settings in your conda installation. They are saved in the file `~/.condarc`\n",
        "\n",
        "```{.bash}\n",
        "cat ~/.condarc\n",
        "```\n",
        "\n",
        "## Create an environment\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "An empty environment called `test_1`:\n",
        "\n",
        "```{.bash}\n",
        "conda create -n test_1\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "You can list all the environments available:\n",
        "\n",
        "```{.bash}\n",
        "conda env list\n",
        "```\n",
        "```\n",
        "> # conda environments:\n",
        "> #\n",
        "> base      *  /home/samuele/miniconda3\n",
        "> test_1       /home/samuele/miniconda3/envs/test_1\n",
        "```\n",
        "\n",
        "## Activate and deactivate\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "To use an environment\n",
        "\n",
        "```{.bash}\n",
        "conda activate test_1\n",
        "```\n",
        "\n",
        "Deactivation happens by\n",
        "\n",
        "```{.bash}\n",
        "conda deactivate\n",
        "```\n",
        "\n",
        "## Manage an environment\n",
        "\n",
        "### Package installation\n",
        "\n",
        "Conda puts together the dependency trees of requested packages to find all compatible dependencies versions.\n",
        "\n",
        "![Figure: A package's dependency tree with required versions on the edges](img/condatree.png)"
      ],
      "id": "a89dcce3"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "To install a specific package in your environment, search it on [anaconda.org](https://anaconda.org):\n",
        "\n",
        "![Figure: search DeSeq2 for R](img/anaconda1.png)\n",
        "\n",
        "---"
      ],
      "id": "dd47b32d"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "![Figure: suggested commands to install the package](img/anaconda2.png){width=10cm}\n",
        "\n",
        ":::{.callout-note title=\"Repositories\"}\n",
        "\n",
        "packages are archived in repositories. Typical ones are `bioconda`, `conda-forge`, `r`, `anaconda`.\n",
        "\n",
        "`conda-forge` packges are often more up-to-date, but a few times show compatibility problems with other packages.\n",
        ":::"
      ],
      "id": "fd9d764e"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "Install a couple of packages in the activated environment - you can always specify a version restriction to each package:\n",
        "\n",
        "```{.bash}\n",
        "conda activate test_1\n",
        "conda install bioconda::bioconductor-deseq2<=1.42.0 conda-forge::r-tidyr=1.3.1\n",
        "```\n",
        "\n",
        ":::{.callout-note}\n",
        "To install two packages, you need more than a hundred installations! Those are all dependencies arising from the **comparison of dependency trees**.\n",
        ":::\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Look for the package `tidyr` in your active environment:\n",
        "\n",
        "```{.bash}\n",
        "conda list | grep tidyr\n",
        "```\n",
        "\n",
        "---"
      ],
      "id": "4d1005ce"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Installation from a list of packages\n",
        "\n",
        "You can export all the packages you have installed over time in your environment:\n",
        "\n",
        "```{.bash}\n",
        "conda env export --from-history > environment.yml\n",
        "```\n",
        "\n",
        "which looks like \n",
        "\n",
        "```{.yaml}\n",
        "name: test_1\n",
        "channels:\n",
        " - bioconda\n",
        " - conda-forge\n",
        " - defaults\n",
        " - r\n",
        "dependencies:\n",
        " - bioconda::bioconductor-deseq2\n",
        " - conda-forge::r-tidyr\n",
        "```"
      ],
      "id": "b9e0618a"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "The same command without `--from-history` will create a very long file with ALL dependencies:\n",
        "\n",
        "```{.yaml}\n",
        "name: test_1\n",
        "channels:\n",
        "  - bioconda\n",
        "  - conda-forge\n",
        "  - defaults\n",
        "  - r\n",
        "dependencies:\n",
        "  - _libgcc_mutex=0.1=conda_forge\n",
        "  - _openmp_mutex=4.5=2_gnu\n",
        "  - _r-mutex=1.0.1=anacondar_1\n",
        "  - argcomplete=3.2.2=pyhd8ed1ab_0\n",
        "  - binutils_impl_linux-64=2.40=hf600244_0\n",
        "```\n",
        "\n",
        "This is guaranteed to work **only on the specific system** where you created the environment!\n",
        "\n",
        "---"
      ],
      "id": "6381e0ae"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "You can use the `yml` file to create an environment:\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "```\n",
        "conda env create -p test_1_from_file -f ./environment.yml\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Environment files are very useful when you want to **share environments with others**, especially when the package list is long."
      ],
      "id": "a58afcd7"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "**Good practice:** You want to install a lot of packages in an environment? Clone it first!\n",
        "If you break something, you still have the old copy.\n",
        "\n",
        "```{.bash}\n",
        "conda create -p test_1_cloned --clone test_1\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "If installations in the cloned environment go fine, then you can remove it \n",
        "\n",
        "```{.bash}\n",
        "conda env remove -n test_1_cloned\n",
        "```\n",
        "\n",
        "and repeat the installations on the original one.\n",
        "\n",
        "---"
      ],
      "id": "56ed67f6"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Useful links for conda:\n",
        "\n",
        "- [Conda cheat sheet](https://docs.conda.io/projects/conda/en/4.6.0/_downloads/52a95608c49671267e40c689e0bc00ca/conda-cheatsheet.pdf) with all the things you can do to manage environments\n",
        "\n",
        "- [Anaconda](http://anaconda.org) where you can search for packages\n",
        "\n",
        "# Running a Job\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Running programs on a computing cluster happens through **jobs**. \n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Learn how to get hold of **computing resources** to run your programs.\n",
        "\n",
        "## What is a job on a HPC\n",
        "\n",
        "A computational task **executed on requested HPC resources** (computing nodes), which are handled by the **queueing system** (SLURM).\n",
        "\n",
        "![](img/Job-on-cluster.png){fig-align=\"center\"}"
      ],
      "id": "33f4c386"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "\n",
        "The command `gnodes` will tell you if there is heavy usage across the computing nodes\n",
        "\n",
        "![Usage of computing nodes. Each node has a name (e.g. cn-1001). The symbols for each node mean running a program (`0`), assigned to an user (`_`) and available (`.`)](./img/gnodes.png)\n",
        "\n",
        "If you want to venture more into checking the queueing status, Moi has done [a great interactive script](https://github.com/MoiColl/cluster_status) in R Shiny for that.\n",
        "\n",
        "---"
      ],
      "id": "83896f6f"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Front-end nodes are limited in memory and power, and **should only be for basic operations** such as\n",
        "\n",
        "- starting a new project\n",
        "\n",
        "- small folders and files management\n",
        "\n",
        "- small software installations\n",
        "\n",
        "and in general you **should not** use them to run computations. This might slow down other users."
      ],
      "id": "50215e60"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "## Interactive jobs\n",
        "\n",
        "Useful to run a **non-repetitive task interactively**\n",
        "\n",
        "Examples: \n",
        "\n",
        "- splitting by chromosome that one `bam` file you just got \n",
        "\n",
        "- open `python`/`R` and do some statistics \n",
        "\n",
        "- compress/decompress multiple files, maybe in parallel\n",
        "\n",
        "Once you exit from the job, anything running into it will stop.\n",
        "\n",
        "---"
      ],
      "id": "1e2249c4"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "To run an interactive job simply use the command\n",
        "\n",
        "```\n",
        "[fe-open-01]$ srun --mem=<GB_of_RAM>g -c <nr_cores> --time=<days-hrs:mins:secs>  --account=<project_name> --pty /bin/bash\n",
        "```\n",
        "\n",
        "For example\n",
        "\n",
        "```\n",
        "[fe-open-01]$ srun --mem=32g -c 2 --time=6:0:0  --account=<project_name> --pty /bin/bash\n",
        "```\n",
        "\n",
        "The **queueing system** makes you wait based on the resources you ask and how busy the nodes are. When you get **assigned a node**, the resources are available. The node name is **shown in the prompt**.\n",
        "\n",
        "```\n",
        "[<username>@s21n32 ~]$\n",
        "```\n",
        "\n",
        "## Batch script (sbatch)\n",
        "\n",
        "Useful to **run a program non-interactively**, usually for longer time than a short interaction. A batch script contains \n",
        "\n",
        "- the desired **resources**\n",
        "- the sequence of **commands** to be executed\n",
        "\n",
        "and\n",
        "\n",
        "- has a filename **without spaces** (forget spaces from now on)\n",
        "- starts with `#!/bin/bash` to know in which language ('bash') the commands are written into"
      ],
      "id": "9e52215d"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "### Example \n",
        "\n",
        "A file called `align.sh` such that:\n",
        "\n",
        "```{.bash}\n",
        "#!/bin/bash\n",
        "#SBATCH --account my_project\n",
        "#SBATCH --cpus-per-task= 8\n",
        "#SBATCH --mem 16g\n",
        "#SBATCH --time 04:00:00\n",
        "\n",
        "#activate environment\n",
        "eval \"$(conda shell.bash hook)\"\n",
        "conda activate ./bam_tools \n",
        "#index the reference file\n",
        "bwa-mem2 index reference/chr2.fa\n",
        "#align data\n",
        "bwa-mem2 -t 8 reference/chr2.fa \\\n",
        "             genomes/S_Korean-2.region.fq.gz \\\n",
        "        | samtools sort \\\n",
        "            -@ 7 \\\n",
        "            -n \\\n",
        "            -O BAM \\\n",
        "        > alignment/S_Korean-2.sorted.bam\n",
        "\n",
        "exit 0\n",
        "```\n",
        "\n",
        "---"
      ],
      "id": "d0823047"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Send the script to the queueing system:\n",
        "\n",
        "```{.bash}\n",
        "sbatch align.sh\n",
        "```\n",
        "```\n",
        "Submitted batch job 33735298\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "Interrogate SLURM about the specific job\n",
        "\n",
        "```{.bash}\n",
        "jobinfo 33735298\n",
        "```\n",
        "\n",
        "```\n",
        ">Name                : align.sh\n",
        ">User                : samuele\n",
        ">Account             : my_project\n",
        ">Partition           : short\n",
        ">Nodes               : s21n43\n",
        ">Cores               : 8\n",
        ">GPUs                : 0\n",
        ">State               : RUNNING\n",
        ">...\n",
        "```"
      ],
      "id": "022e14a1"
    },
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "or about all the queued jobs\n",
        "\n",
        "```{.bash}\n",
        "squeue -u <username>\n",
        "```\n",
        "\n",
        "```\n",
        ">JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)\n",
        ">33735928     short align.sh  samuele  R       1:12      1 s21n43\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "If you change your mind and want to cancel a job:\n",
        "\n",
        "```{.bash}\n",
        "scancel 33735928\n",
        "```\n",
        "\n",
        "---"
      ],
      "id": "d1a6297f"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "To observe in real time the latest output of the command in the job, you can refresh the last lines of the log file for the specific job:\n",
        "\n",
        "```{.bash}\n",
        "watch tail align.sh-33735928.out\n",
        "```\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "To look at the whole file, you can run at any time\n",
        "\n",
        "```{.bash}\n",
        "less -S align.sh-33735928.out\n",
        "```\n",
        "\n",
        "This can be useful for debugging, when for example a command gives an error and the job interrupts.\n",
        "\n",
        "\n",
        "\n",
        "## Other ways of running jobs\n",
        "\n",
        "Beyond `sbatch`, you can use other tools for workflows which are\n",
        "\n",
        "- **modular and composable**: sequences of commands can be applied in various contexts, composed together in the desired ordering\n",
        "- **scalable and parallel** handling many sequences of operations parallelly or interdependently\n",
        "- **flexible** where repetitive operations can be automatized over multiple applications"
      ],
      "id": "347a8710"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "---\n",
        "\n",
        "Some workflow tools:\n",
        "\n",
        "- [Snakemake](https://snakemake.readthedocs.io/en/stable/)\n",
        "- [NextFlow](https://www.nextflow.io/)\n",
        "- [Gwf](https://gwf.app/)\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "`Gwf` has an easy `python` syntax instead of its own language to write workflows.\n",
        "\n",
        "&nbsp;\n",
        "\n",
        "You need to know some basic `python` to use `Gwf`, but it is worth the effort."
      ],
      "id": "38f7c852"
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}